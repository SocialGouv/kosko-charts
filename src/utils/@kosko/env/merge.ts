//

/* eslint-disable @typescript-eslint/no-unsafe-return */
/* eslint-disable @typescript-eslint/no-explicit-any */

// HACK(douglasduteil): wrap internal @kosko/env/dist/merge.js
// As the paths API is internal (not exported) in @kosko/env
// This is code will wrap and expose it for our needs.
// see https://github.com/tommy351/kosko/blob/%40kosko/env%401.0.1/packages/env/src/merge.ts#L12-L18
// since https://github.com/tommy351/kosko/commit/6353c8559994b826c8056ad03a35332e88b9134a

// https://stackoverflow.com/a/48769843
type UnionToIntersection<T> = (T extends any ? (k: T) => void : never) extends (
  k: infer I
) => void
  ? I
  : never;

export const merge: <T extends any[]>(
  ...data: T
) => UnionToIntersection<T[number]> =
  // eslint-disable-next-line @typescript-eslint/no-require-imports, @typescript-eslint/no-var-requires
  require("@kosko/env/dist/merge.js").merge;
