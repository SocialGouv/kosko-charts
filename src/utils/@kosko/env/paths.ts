//

import { join } from "path";

// HACK(douglasduteil): wrap internal @kosko/env/dist/paths.js
// As the paths API is internal (not exported) in @kosko/env
// This is code will wrap and expose it for our needs.
// see https://github.com/tommy351/kosko/blob/%40kosko/env%400.5.2/packages/env/src/paths.ts#L9-L13
// since https://github.com/tommy351/kosko/commit/6353c8559994b826c8056ad03a35332e88b9134a

export const formatPath: (
  path: string,
  data: Record<string, unknown>
) => string =
  // eslint-disable-next-line @typescript-eslint/no-require-imports, @typescript-eslint/no-var-requires
  require(join(require.resolve("@kosko/env"), "../paths.js")).formatPath;
