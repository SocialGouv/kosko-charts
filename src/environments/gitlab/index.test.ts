import env from "./index";

test.each([
  ["because of missing variables", {}],
  [
    "because of empty strings",
    {
      CI_COMMIT_TAG: "",
      CI_ENVIRONMENT_NAME: "",
      CI_ENVIRONMENT_SLUG: "",
      CI_PROJECT_NAME: "",
      CI_PROJECT_PATH_SLUG: "",
      KUBE_INGRESS_BASE_DOMAIN: "",
      KUBE_NAMESPACE: "",
      PRODUCTION: "",
    },
  ],
])("should throw %s", (_: string, testEnv?: NodeJS.ProcessEnv) => {
  expect(() => env(testEnv)).toThrowErrorMatchingSnapshot();
});

const validEnv = {
  CI_ENVIRONMENT_NAME: "fabrique-dev",
  CI_ENVIRONMENT_SLUG: "my-test",
  CI_PROJECT_NAME: "sample",
  CI_PROJECT_PATH_SLUG: "socialgouv-sample",
  CI_REGISTRY_IMAGE: "registry.gitlab.factory.social.gouv.fr/socialgouv/sample",
  KUBE_INGRESS_BASE_DOMAIN: "dev.fabrique.social.gouv.fr",
  KUBE_NAMESPACE: "sample-42-my-test",
};
test.each([
  ["the gitlab global env", { ...validEnv }],
  [
    "tagged gitlab global env",
    {
      ...validEnv,
      CI_COMMIT_TAG: "vX.Y.Z",
    },
  ],
  [
    "production gitlab global env",
    {
      ...validEnv,
      CI_COMMIT_TAG: "vX.Y.Z",
      PRODUCTION: "true",
    },
  ],
])("should return %s", (_: string, testEnv?: NodeJS.ProcessEnv) => {
  expect(env(testEnv)).toMatchSnapshot();
});
