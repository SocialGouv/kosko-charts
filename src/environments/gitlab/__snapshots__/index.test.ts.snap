// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return production gitlab global env 1`] = `
Object {
  "annotations": Object {
    "app.gitlab.com/app": "socialgouv-sample",
    "app.gitlab.com/env": "my-test",
    "app.gitlab.com/env.name": "fabrique-dev",
  },
  "domain": "dev.fabrique.social.gouv.fr",
  "labels": Object {
    "application": "sample",
    "owner": "sample",
    "team": "sample",
  },
  "namespace": Object {
    "name": "sample",
  },
  "subdomain": "sample",
}
`;

exports[`should return tagged gitlab global env 1`] = `
Object {
  "annotations": Object {
    "app.gitlab.com/app": "socialgouv-sample",
    "app.gitlab.com/env": "my-test",
    "app.gitlab.com/env.name": "fabrique-dev",
  },
  "domain": "dev.fabrique.social.gouv.fr",
  "labels": Object {
    "application": "vX-Y-Z-sample",
    "owner": "sample",
    "team": "sample",
  },
  "namespace": Object {
    "name": "sample-42-my-test",
  },
  "subdomain": "vX-Y-Z-sample",
}
`;

exports[`should return the gitlab global env 1`] = `
Object {
  "annotations": Object {
    "app.gitlab.com/app": "socialgouv-sample",
    "app.gitlab.com/env": "my-test",
    "app.gitlab.com/env.name": "fabrique-dev",
  },
  "domain": "dev.fabrique.social.gouv.fr",
  "labels": Object {
    "application": "my-test-sample",
    "owner": "sample",
    "team": "sample",
  },
  "namespace": Object {
    "name": "sample-42-my-test",
  },
  "subdomain": "my-test-sample",
}
`;

exports[`should throw because of empty strings 1`] = `
"BadArgument:required property \\"CI_ENVIRONMENT_NAME\\"
└─ cannot decode \\"\\", should be NonEmptyString
required property \\"CI_ENVIRONMENT_SLUG\\"
└─ cannot decode \\"\\", should be NonEmptyString
required property \\"CI_PROJECT_NAME\\"
└─ cannot decode \\"\\", should be NonEmptyString
required property \\"CI_PROJECT_PATH_SLUG\\"
└─ cannot decode \\"\\", should be NonEmptyString
required property \\"KUBE_INGRESS_BASE_DOMAIN\\"
└─ cannot decode \\"\\", should be NonEmptyString
required property \\"KUBE_NAMESPACE\\"
└─ cannot decode \\"\\", should be NonEmptyString
optional property \\"CI_COMMIT_TAG\\"
└─ cannot decode \\"\\", should be NonEmptyString"
`;

exports[`should throw because of missing variables 1`] = `
"BadArgument:required property \\"CI_ENVIRONMENT_NAME\\"
└─ cannot decode undefined, should be string
required property \\"CI_ENVIRONMENT_SLUG\\"
└─ cannot decode undefined, should be string
required property \\"CI_PROJECT_NAME\\"
└─ cannot decode undefined, should be string
required property \\"CI_PROJECT_PATH_SLUG\\"
└─ cannot decode undefined, should be string
required property \\"KUBE_INGRESS_BASE_DOMAIN\\"
└─ cannot decode undefined, should be string
required property \\"KUBE_NAMESPACE\\"
└─ cannot decode undefined, should be string"
`;
